File: mavsim_matlab/chap7/mavsim_chap7.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% mavsim_matlab 
%     - Chapter 7 assignment for Beard & McLain, PUP, 2012
%     - Update history:  
%         2/18/2019 - RWB
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
run('../parameters/simulation_parameters')  % load SIM: simulation parameters
run('../parameters/aerosonde_parameters')  % load MAV: aircraft parameters
run('../parameters/sensor_parameters')  % load SENSOR: sensor parameters

% initialize the mav viewer
addpath('../chap2'); mav_view = mav_viewer();  
addpath('../chap3'); data_view = data_viewer();

% initialize the video writer
VIDEO = 0;  % 1 means write video, 0 means don't write video
if VIDEO==1, video=video_writer('chap7_video.avi', SIM.ts_video); end

% initialize elements of the architecture
addpath('../chap4'); wind = wind_simulation(SIM.ts_simulation);
addpath('../chap6'); ctrl = autopilot(SIM.ts_simulation);
addpath('../chap7'); mav = mav_dynamics(SIM.ts_simulation, MAV);

addpath('../message_types'); commands = msg_autopilot();
addpath('../tools');

% arguments to signals are amplitude, frequency, start_time, dc_offset
Va_command = signals(3, 0.01, 2, 25);
h_command = signals(10, 0.02, 0, 100);
chi_command = signals(45*pi/180, 0.015, 5, 180*pi/180);

% initialize the simulation time
sim_time = SIM.start_time;

% main simulation loop
disp('Type CTRL-C to exit');
while sim_time < SIM.end_time
    %-------controller-------------
    measurements = mav.sensors(MAV, SENSOR);
    estimated_state = mav.true_state;  % uses true states in the control
    commands.airspeed_command = Va_command.square(sim_time);
    commands.course_command = chi_command.square(sim_time);
    commands.altitude_command = h_command.square(sim_time);
    [delta, commanded_state] = ctrl.update(commands, estimated_state);
    
    %-------physical system-------------
    current_wind = wind.update();
    mav.update(delta, current_wind, MAV);
    
    %-------update viewer-------------
    mav_view.update(mav.true_state);       % plot body of MAV
    data_view.update(mav.true_state,...    % true states
                     estimated_state,...       % estimated states
                     commanded_state,...       % commmanded states
                     SIM.ts_simulation); 
    if VIDEO==1, video.update(sim_time); end

    %-------increment time-------------
    sim_time = sim_time + SIM.ts_simulation;
end

if VIDEO==1, video.close(); end

File: mavsim_python/launch_files/chap07/mavsim_chap7.py
"""
mavsim_python
    - Chapter 7 assignment for Beard & McLain, PUP, 2012
    - Last Update:
        2/18/2020 - RWB
        1/5/2023 - David L. Christiansen
        7/13/2023 - RWB
"""
import os, sys
# insert parent directory at beginning of python search path
from pathlib import Path
sys.path.insert(0,os.fspath(Path(__file__).parents[1]))
# use QuitListener for Linux or PC <- doesn't work on Mac
#from tools.quit_listener import QuitListener
import numpy as np
import parameters.simulation_parameters as SIM
import pyqtgraph as pg
from tools.signals import Signals
from models.mav_dynamics_sensors import MavDynamics
from models.wind_simulation import WindSimulation
from control.autopilot import Autopilot
from viewers.mav_viewer import MavViewer
from viewers.data_viewer import DataViewer
from viewers.sensor_viewer import SensorViewer

#quitter = QuitListener()

VIDEO = False
DATA_PLOTS = False
SENSOR_PLOTS = True
ANIMATION = True
SAVE_PLOT_IMAGE = False

# video initialization
if VIDEO is True:
    from viewers.video_writer import VideoWriter
    video = VideoWriter(video_name="chap7_video.avi",
                        bounding_box=(0, 0, 1000, 1000),
                        output_rate=SIM.ts_video)

#initialize the visualization
if ANIMATION or DATA_PLOTS or SENSOR_PLOTS:
    app = pg.QtWidgets.QApplication([]) # use the same main process for Qt applications
if ANIMATION:
    mav_view = MavViewer(app=app)  # initialize the mav viewer
if DATA_PLOTS:
    data_view = DataViewer(app=app,dt=SIM.ts_simulation, plot_period=SIM.ts_plot_refresh, 
                           data_recording_period=SIM.ts_plot_record_data, time_window_length=30)
if SENSOR_PLOTS:
    sensor_view = SensorViewer(app=app,dt=SIM.ts_simulation, plot_period=SIM.ts_plot_refresh, 
                           data_recording_period=SIM.ts_plot_record_data, time_window_length=30)


# initialize elements of the architecture
wind = WindSimulation(SIM.ts_simulation)
mav = MavDynamics(SIM.ts_simulation)
autopilot = Autopilot(SIM.ts_simulation)

# autopilot commands
from message_types.msg_autopilot import MsgAutopilot
commands = MsgAutopilot()
Va_command = Signals(dc_offset=25.0,
                     amplitude=3.0,
                     start_time=2.0,
                     frequency=0.01)
altitude_command = Signals(dc_offset=100.0,
                           amplitude=10.0,
                           start_time=0.0,
                           frequency=0.02)
course_command = Signals(dc_offset=np.radians(180),
                         amplitude=np.radians(45),
                         start_time=5.0,
                         frequency=0.015)

# initialize the simulation time
sim_time = SIM.start_time
end_time = 100

# main simulation loop
print("Press 'Esc' to exit...")
while sim_time < end_time:

    # -------autopilot commands-------------
    commands.airspeed_command = Va_command.square(sim_time)
    commands.course_command = course_command.square(sim_time)
    commands.altitude_command = altitude_command.square(sim_time)

    # -------autopilot-------------
    measurements = mav.sensors()  # get sensor measurements
    estimated_state = mav.true_state  # uses true states in the control
    delta, commanded_state = autopilot.update(commands, estimated_state)

    # -------physical system-------------
    current_wind = wind.update()  # get the new wind vector
    mav.update(delta, current_wind)  # propagate the MAV dynamics

    # -------update viewer-------------
    if ANIMATION:
        mav_view.update(mav.true_state)  # plot body of MAV
    if DATA_PLOTS:
        plot_time = sim_time
        data_view.update(mav.true_state,  # true states
                            None,  # estimated states
                            commanded_state,  # commanded states
                            delta)  # inputs to aircraft
    if SENSOR_PLOTS:
        sensor_view.update(measurements)

    if ANIMATION or DATA_PLOTS or SENSOR_PLOTS:
        app.processEvents()

    # sensor_view.update(mav.sensors(),  # sensor values
    #                    SIM.ts_simulation)
    if VIDEO is True:
        video.update(sim_time)
        
    # -------Check to Quit the Loop-------
    if quitter.check_quit():
        break

    # -------increment time-------------
    sim_time += SIM.ts_simulation

# Save an Image of the Plot
if SAVE_PLOT_IMAGE:
    if DATA_PLOTS:
        data_view.save_plot_image("ch7_data_plot")
    if SENSOR_PLOTS:
        sensor_view.save_plot_image("ch7_sensor_plot")

if VIDEO is True:
    video.close()
